pkg/kubelet/pod_workers.go

```
@@ -139,77 +364,233 @@ type podWorkers struct {
+			status.terminatingAt = now
+			becameTerminating = true
+		case options.UpdateType == kubetypes.SyncPodKill:
+			if options.KillPodOptions != nil && options.KillPodOptions.Evict {
+				klog.V(4).InfoS("Pod is being evicted by the kubelet, begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+				status.evicted = true
+			} else {
+				klog.V(4).InfoS("Pod is being removed by the kubelet, begin teardown", "pod", klog.KObj(pod), "podUID", pod.UID)
+			}
+			status.terminatingAt = now
+			becameTerminating = true
+		}
+	}
+
+	// once a pod is terminating, all updates are kills and the grace period can only decrease
+	var workType PodWorkType
+	var wasGracePeriodShortened bool
+	switch {
+	case status.IsTerminated():
+		workType = TerminatedPodWork
+
+		if options.KillPodOptions != nil {
+			if ch := options.KillPodOptions.CompletedCh; ch != nil {
+				close(ch)
+			}
+		}
+		options.KillPodOptions = nil
+
+	case status.IsTerminating():
+		workType = TerminatingPodWork
+		if options.KillPodOptions == nil {
+			options.KillPodOptions = &KillPodOptions{}
+		}
+
+		if ch := options.KillPodOptions.CompletedCh; ch != nil {
+			status.notifyPostTerminating = append(status.notifyPostTerminating, ch)
+		}
+
+		gracePeriod, gracePeriodShortened := calculateEffectiveGracePeriod(status, pod, options.KillPodOptions)
+
+		wasGracePeriodShortened = gracePeriodShortened
+		status.gracePeriod = gracePeriod
+		// always set the grace period for syncTerminatingPod so we don't have to recalculate,
+		// will never be zero.
+		options.KillPodOptions.PodTerminationGracePeriodSecondsOverride = &gracePeriod
+
+		// if a static pod comes through, start tracking it explicitly (cleared by the pod worker loop)
+		if kubelettypes.IsStaticPod(pod) {
+			p.terminatingStaticPodFullnames[kubecontainer.GetPodFullName(pod)] = struct{}{}
+		}
+
+	default:
+		workType = SyncPodWork
+
+		// completely ignore KillPodOptions
```

When could this happen? Should we return an error instead?

---

We could simply have another public method besides UpdatePod which is "kill pod"

---

Let's at least add comment to point out that this is not a valid case.